src/test/fio/fio_ceph_objectstore.cc

```diff
@@ -0,0 +1,403 @@
+}
+
+Job::~Job()
+{
+  if (unlink) {
+    ObjectStore::Transaction t;
+    // remove our objects
+    for (auto& obj : objects) {
+      t.remove(obj.coll.cid, obj.oid);
+    }
+    // remove our collections
+    for (auto& coll : collections) {
+      t.remove_collection(coll.cid);
+    }
+    ObjectStore::Sequencer sequencer("job cleanup");
+    int r = engine->os->apply_transaction(&sequencer, std::move(t));
+    if (r)
+      derr << "job cleanup failed with " << cpp_strerror(-r) << dendl;
+  }
+}
+
+
+int fio_ceph_os_setup(thread_data* td)
+{
+  // if there are multiple jobs, they must run in the same process against a
+  // single instance of the ObjectStore. explicitly disable fio's default
+  // job-per-process configuration
+  td->o.use_thread = 1;
+
+  try {
+    // get or create the global Engine instance
+    auto engine = Engine::get_instance(td);
+    // create a Job for this thread
+    td->io_ops_data = new Job(engine, td);
+  } catch (std::exception& e) {
+    std::cerr << "setup failed with " << e.what() << std::endl;
+    return -1;
+  }
+  return 0;
+}
+
+void fio_ceph_os_cleanup(thread_data* td)
+{
+  auto job = static_cast<Job*>(td->io_ops_data);
+  td->io_ops_data = nullptr;
+  delete job;
+}
+
+
+io_u* fio_ceph_os_event(thread_data* td, int event)
+{
+  // return the requested event from fio_ceph_os_getevents()
+  auto job = static_cast<Job*>(td->io_ops_data);
+  return job->events[event];
+}
+
+int fio_ceph_os_getevents(thread_data* td, unsigned int min,
+                          unsigned int max, const timespec* t)
+{
+  auto job = static_cast<Job*>(td->io_ops_data);
+  unsigned int events = 0;
+  io_u* u;
+  unsigned int i;
+
+  // loop through inflight ios until we find 'min' completions
+  do {
+    io_u_qiter(&td->io_u_all, u, i) {
+      if (!(u->flags & IO_U_F_FLIGHT))
+        continue;
+
+      if (u->engine_data) {
+        u->engine_data = nullptr;
+        job->events[events] = u;
+        events++;
+      }
+    }
+    if (events >= min)
+      break;
+    usleep(100);
+  } while (1);
+
+  return events;
+}
+
+/// completion context for ObjectStore::queue_transaction()
+class UnitComplete : public Context {
+  io_u* u;
+ public:
+  UnitComplete(io_u* u) : u(u) {}
+  void finish(int r) {
+    // mark the pointer to indicate completion for fio_ceph_os_getevents()
+    u->engine_data = reinterpret_cast<void*>(1ull);
+  }
+};
+
+int fio_ceph_os_queue(thread_data* td, io_u* u)
+{
+  fio_ro_check(td, u);
+
+  auto job = static_cast<Job*>(td->io_ops_data);
+  auto& object = job->objects[u->file->engine_data];
+  auto& coll = object.coll;
+  auto& os = job->engine->os;
+
+  if (u->ddir == DDIR_WRITE) {
+    bufferlist bl;
+    bl.push_back(buffer::create_static(u->xfer_buflen,
+                                       static_cast<char*>(u->xfer_buf)));
+    // enqueue a write transaction on the collection's sequencer
+    ObjectStore::Transaction t;
+    t.write(coll.cid, object.oid, u->offset, u->xfer_buflen, bl);
```

Comment:
You may consider adding CEPH_OSD_OP_FLAG_FADVISE_WILLNEED flag for read/write modes to provide written data caching.